sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala

```diff
@@ -70,437 +75,362 @@ object AggregateUtils {
+              aggregatedBatches.forEach(b => batchesToConcat += b)
+              aggregatedBatches.clear()
+              val batch = concatenateAndMerge(batchesToConcat)
+              // batch does not need to be marked spillable since it is the last and only batch
+              // and will be immediately retrieved on the next() call.
+              aggregatedBatches.add(batch)
+            }
           }
+          return
         }
       }
     }
   }
-}
 
-/**
- * GpuHashAggregateExec - is the GPU version of HashAggregateExec, with some major differences:
- * - it doesn't support spilling to disk
- * - it doesn't support strings in the grouping key
- * - it doesn't support count(col1, col2, ..., colN)
- * - it doesn't support distinct
- * @param requiredChildDistributionExpressions this is unchanged by the GPU. It is used in
- *                                             EnsureRequirements to be able to add shuffle nodes
- * @param groupingExpressions The expressions that, when applied to the input batch, return the
- *                            grouping key
- * @param aggregateExpressions The GpuAggregateExpression instances for this node
- * @param aggregateAttributes References to each GpuAggregateExpression (attribute references)
- * @param resultExpressions the expected output expression of this hash aggregate (which this
- *                          node should project)
- * @param child incoming plan (where we get input columns from)
- */
-case class GpuHashAggregateExec(
-    requiredChildDistributionExpressions: Option[Seq[Expression]],
-    groupingExpressions: Seq[Expression],
-    aggregateExpressions: Seq[GpuAggregateExpression],
-    aggregateAttributes: Seq[Attribute],
-    resultExpressions: Seq[NamedExpression],
-    child: SparkPlan) extends UnaryExecNode with GpuExec with Arm {
+  /**
+   * Perform a single pass over the aggregated batches attempting to merge adjacent batches.
+   * @return true if at least one merge operation occurred
+   */
+  private def mergePass(): Boolean = {
+    val batchesToConcat: mutable.ArrayBuffer[LazySpillableColumnarBatch] = mutable.ArrayBuffer.empty
+    var wasBatchMerged = false
+    var concatSize: Long = 0L
+    var batchesLeftInPass = aggregatedBatches.size()
+
+    while (batchesLeftInPass > 0) {
+      closeOnExcept(batchesToConcat) { _ =>
+        var isConcatSearchFinished = false
+        // Although tempting to allow the pass to "wrap around" and pick up batches previously
+        // merged in this pass, it's avoided to prevent changing the order of aggregated batches.
+        while (batchesLeftInPass > 0 && !isConcatSearchFinished) {
+          val candidate = aggregatedBatches.getFirst
+          val potentialSize = concatSize + candidate.deviceMemorySize
+          isConcatSearchFinished = concatSize > 0 && potentialSize > targetBatchSize
+          if (!isConcatSearchFinished) {
```

Comment:
Yes, this would occur when there's exactly one batch after the concat search is finished.  For example, we pick up some large first batch unconditionally, because concatSize == 0.  Then on the next batch we find that adding it would exceed the target batch size, so we mark the search finished and exit the loop without adding it to the concat list.  That's when we hit the else clause, because batchesToConcat.length <= 1 and there's no point in trying to concatenate a single batch.

Reply:
@jlowe clarified this further as I still was confused (my bad). I had missed that the condition in line 266 was `batchesToConcat.length > 1`  not `batchesToConcat.length > 0`.